File: three.py
#!/usr/bin/python
# Three Card Game Paradox
# One of three cards has a prize
# The player selects a card
# Then a wrong card from the other 2 cards is removed randomly
# The player can stick with original selection or change to the other card
# People tend to stick, however changing is better
# This program simulates a number of these games to prove this
#
# by Conor Gilmer

import random
stick = 0
twist = 0
stickpc = 0.0
twistpc = 0.0
games = 1000
cards = ['A', 'B', 'C']

def iteration(limit):
	winning = random.randint(1,limit)
	chosen = random.randint(1,limit)

	remove = random.randint(1,limit)
	while remove == chosen or remove == winning:
		remove = random.randint(1,limit)
	results = "Winning = " + cards[winning-1] + " | Chosen = " + cards[chosen-1] + " | removed = " + cards[remove-1] 
	if winning == chosen:
		print results + " - Verdict : Sticking Wins"
		return 1  
	else:
		print results + " - Verdict : Changing Wins"
		return 0


print "Three Card Game Paradox"
print "Playing %d times" % games

for i in range(1, games+1):
	if iteration(3) == 1:
		stick = stick +1
	else:
		twist = twist +1

stickpc = (float(stick)/float(games)) * 100
twistpc = (float(twist)/float(games)) * 100

print "Results:"
s ="Sticking wins " + str(stick) + " times ("+ str(stickpc) + "% )"
print s
s ="Changing wins " + str(twist) + " times ("+ str(twistpc) + "% )"
print s

if stick > twist:
	print "Sticking is better!"
elif stick == twist:
	print "Sticking and twisting are the same!"
else:
	print "Changing is better!"

print "The End"

File: three.pl
#!/usr/bin/perl -w
# ---------------------------------- #
# Three Card Game paradox three.pl
# by Conor Gilmer
# ---------------------------------- # 

use strict;
use warnings;

my $winning;		# var for winning card
my $chosen;		# var for chosen card
my $remove;		# var for random wrong card to be removed
my $twist	= 0;	# counter for number of winning twists
my $stick	= 0;	# counter for number of winning sticks;
my $x;
my $numA 	= 0;
my $numB 	= 0;
my $numC 	= 0;
my $randomized;

my @cards       = ('A', 'B', 'C');	# array of cards


sub randomise
{ my $var = shift;
  my $randnum =0;
  $randnum = int(rand($var));
  if ($randnum == 0) {
	$numA = $numA +1;}
  elsif ($randnum == 1) {
	$numB = $numB +1;}
  elsif ($randnum == 2) {
	$numC = $numC +1;}
  return $randnum;
}

sub run
{
	#print ("Three cards ( @cards )\n");
	#winning card picked (random)
	#$winning = int(rand(3)); #cast the random number to an integer
	$winning = randomise(3);
	#card selected (random)
	#$chosen = int(rand(3));
	$chosen = randomise(3);
	
	#remove one wrong card
	do {
		#$remove = int(rand(3));
		$remove = randomise(3);
	} while (($cards[$remove] eq $cards[$chosen]) || ($cards[$remove] eq $cards[$winning]) );
	
	print("Winning $cards[$winning] | Choose $cards[$chosen] | Removed $cards[$remove]\n");

	#stick or twist
	print("Winning Card is $cards[$winning]");
	if ($winning == $chosen) 	{
		print (" - Sticking wins\n");
		$stick++;
	}
	else {
		print (" - Twisting wins\n");
		$twist++;
	}
}

sub main
{	my $games = shift;
	my $percentageS;
	my $percentageT;
	print ("\n");
	print ("Three Card Game Paradox \n");
	# run 1000 times
	for (my $i=1; $i<($games+1); $i++)
	{
	   print "Game: $i\n";
	   run();
	}
	print("\nGame played $games times\n");
	$percentageS = 100 * ($stick/$games);
	$percentageT = 100 * ($twist/$games);
	print("Results:\n");
	print("Sticking wins $stick ($percentageS %)times \nTwisting wins $twist ($percentageT %) times\n");
	print("The End\n");

	print("How Random is rand(x)\n");
	$randomized = $numA + $numB + $numC;
	print("rand(x) called $randomized times\n");
	print("A chosen $numA times(".calPC($randomized,$numA, 2)."%), B chosen $numB times(".calPC($randomized,$numB,2)."%), C chosen $numC times(".calPC($randomized,$numC,2)."%).\n");

}

# calculate the percentage
sub calPC
{
	my $total= shift;
	my $amt  = shift;
	my $places = shift;
	return  sprintf("%.".$places."f",(($amt/$total) *100));
}

# check if the passed parameter is a number
sub is_integer {
   defined $_[0] && $_[0] =~ /^[+-]?\d+$/;
}



#get input argument (number of times to run)
$x = shift(@ARGV);

# run main value passed or default 1000 times
if (($x ne "") && (is_integer($x))) {
	main($x);
} else	{
	main(1000);
}
